### PR TITLE
Implement Region for smir

### PR DESCRIPTION
**Pull request by ouz-a**
Adds Region and it's relevant types to smir and covers them with stable implementation

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
Please rebase the PR

File: compiler/rustc_smir/src/stable_mir/fold.rs [hunk 1]
```diff
@@ -106,6 +106,12 @@ impl Foldable for GenericArgs {
     }
 }
 
+impl Foldable for Region {
+    fn super_fold<V: Folder>(&self, _folder: &mut V) -> ControlFlow<V::Break, Self> {
+        ControlFlow::Continue(self.clone())
+    }
+}
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/fold.rs:113]: Regions are in the same category as `Constant` and `Ty`, so you need to add a new method to `Folder` and `Visitor` for them, and then call it from here.

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 2]
```diff
@@ -34,7 +34,59 @@ pub struct Const {
...
+#[derive(Debug, Clone)]
+pub enum RegionKind {
+    ReEarlyBound(EarlyBoundRegion),
+    ReLateBound(DebruijnIndex, BoundRegion),
+    ReFree(FreeRegion),
+    ReStatic,
+    ReVar(RegionVid),
+    RePlaceholder(Placeholder<BoundRegion>),
+    ReErased,
+    ReError(ErrorGuaranteed),
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/ty.rs:52]: we do not forward errors to SMIR in Ty and Const, so we don't need to do it here either.

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 3]
```diff
@@ -34,7 +34,59 @@ pub struct Const {
...
+#[derive(Debug, Clone)]
+pub struct Region {
+    pub kind: RegionKind,
+}
+
+#[derive(Debug, Clone)]
+pub enum RegionKind {
+    ReEarlyBound(EarlyBoundRegion),
+    ReLateBound(DebruijnIndex, BoundRegion),
+    ReFree(FreeRegion),
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/ty.rs:47]: Not sure we can encounter this one

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 4]
```diff
@@ -34,7 +34,59 @@ pub struct Const {
...
+    ReStatic,
+    ReVar(RegionVid),
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/ty.rs:49]: We don't emit type variables in SMIR (we don't encounter them), so we should also not emit region variables

File: compiler/rustc_smir/src/stable_mir/fold.rs (hunk 1 above)
Comment (ouz-a) [reply compiler/rustc_smir/src/stable_mir/fold.rs]: I added them, but not entirely sure with the results 😅

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 5]
```diff
@@ -1500,9 +1501,39 @@ impl<'tcx> Stable<'tcx> for ty::ImplPolarity {
...
+                RegionKind::RePlaceholder(stable_mir::ty::Placeholder {
+                    universe: place_holder.universe.as_u32(),
+                    bound: BoundRegion {
+                        var: place_holder.bound.var.as_u32(),
+                        kind: place_holder.bound.kind.stable(tables),
+                    },
+                })
+            }
+            ty::ReErased => RegionKind::ReErased,
+            _ => unimplemented!(),
```
Comment (oli-obk) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:1535]:
```suggestion
            _ => unreachable!("{self:?}"),
```

`unimplemented!` suggests there's something to do here.

File: compiler/stable_mir/src/fold.rs [hunk 6]
```diff
@@ -15,6 +18,9 @@ pub trait Folder: Sized {
     fn fold_const(&mut self, c: &Const) -> ControlFlow<Self::Break, Const> {
         c.super_fold(self)
     }
+    fn visit_reg(&mut self, reg: &Region) -> ControlFlow<Self::Break, Region> {
```
Comment (oli-obk) [inline compiler/stable_mir/src/fold.rs:21]: lol uhm... I messed up here. Please use `fold_reg` and change `visit_ty` to `fold_ty` while you're at it.

File: compiler/stable_mir/src/fold.rs [hunk 7]
```diff
@@ -106,6 +112,43 @@ impl Foldable for GenericArgs {
...
+
+impl Foldable for BoundRegionKind {
+    fn super_fold<V: Folder>(&self, _folder: &mut V) -> ControlFlow<V::Break, Self> {
+        match self {
+            BoundRegionKind::BrAnon => ControlFlow::Continue(self.clone()),
+            BoundRegionKind::BrNamed(_, _) => ControlFlow::Continue(self.clone()),
+            BoundRegionKind::BrEnv => ControlFlow::Continue(self.clone()),
+        }
+    }
+}
```
Comment (oli-obk) [inline compiler/stable_mir/src/fold.rs:150]: We only need impls for types that contain regions, types or constants. Everything else can just be cloned at the call site.

Comment (oli-obk) [reply compiler/stable_mir/src/fold.rs:150]: The reason we only need it for those 3 types is that these are the only ones where a `Folder` can have any effect, so if we know there is no effect, we can just skip the entire thing.

File: compiler/stable_mir/src/visitor.rs [hunk 8]
```diff
@@ -101,6 +107,39 @@ impl Visitable for GenericArgs {
...
+
+impl Visitable for BoundRegionKind {
+    fn super_visit<V: Visitor>(&self, _visitor: &mut V) -> ControlFlow<V::Break> {
+        match self {
+            BoundRegionKind::BrAnon => ControlFlow::Continue(()),
+            BoundRegionKind::BrNamed(_, _) => ControlFlow::Continue(()),
+            BoundRegionKind::BrEnv => ControlFlow::Continue(()),
+        }
+    }
+}
```
Comment (oli-obk) [inline compiler/stable_mir/src/visitor.rs:141]: same thing, these aren't needed, regions have nothing internal to visit, they have no nested regions, types or consts

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Pull request merged**